src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/UIPropertyDataProducer.cs

```diff
@@ -106,8 +106,15 @@ public static IEntityValue CreateUIPropertyValue(IQueryExecutionContext queryExe
 
             if (requestedProperties.VisibilityCondition)
             {
-                string? visibilityCondition = property.GetMetadataValueOrNull("VisibilityCondition");
-                newUIProperty.VisibilityCondition = visibilityCondition ?? string.Empty;
+                if (!property.Visible)
+                {
+                    newUIProperty.VisibilityCondition = "false";
```

Comment:
I don't think setting this visibility condition is necessary.

A design for this was proposed in https://github.com/dotnet/project-system/issues/7251.

Can we populate the `IUIProperty.IsVisible` property instead? That was added to the project model in https://devdiv.visualstudio.com/DevDiv/_git/CPS/pullrequest/325905

Using `IsVisible` is a clearer way to express to the UI that the property must never be displayed. We can then completely remove these properties from the set of properties for which we evaluate visibility. In the UI, a property that is hidden due to its visibility condition still has several resources allocated for it. I prefer using `IsVisible` to indicate that, rather than interpreting this expression as a constant.

Reply:
> We can then completely remove these properties ...
@drewnoakes exactly which properties are you referring to? the ones on the back end?

Reply:
Ah, I think I was getting that CPS change confused with the `IsReadonly` change.

So on the front-end we would first check `IsVisible`, and only if that is `true` would we evaluate the `VisibilityCondition` (if any)?

My concern is that we now have two properties in the project model (`IsVisible` and `VisibilityCondition`) with overlapping functionality. We can't avoid that on the back end because `BaseProperty.Visible` is already part of the public API, but on the front-end I could imagine using only `VisibilityCondition`:

- If it is not set the property is always visible.
- If it is set to a constant `true` it is always visible.
- If it is set to a constant `false` it is always invisible.

In these cases you would never need to re-evaluate.

That being said I don't feel strongly about it.

Reply:
I tried to make this change but as yet there doesn't seem to be a version of the Microsoft.VisualStudio.ProjectSystem.Query package that includes the `IsVisible` property.

Reply:
> @drewnoakes exactly which properties are you referring to? the ones on the back end?

@sawesonga sorry, I'll be more explicit.

The UI will need all properties, both those which are always non-visible, and those which are visible (with optional conditionality on that visibility).

When a property is conditionally hidden, the UI still allocates UI elements for it. These elements are just not displayed.

For the case where a property is never going to be displayed, we can avoid creating those elements at all. That is what I meant by "completely remove".

> So on the front-end we would first check IsVisible, and only if that is true would we evaluate the VisibilityCondition (if any)?

Sort of. The filtering happens at different layers. The process would be:

1. Get all properties into the `PropertyContext`
2. Create a view of properties to bind to the UI's `ItemsControl`, filtered to those where `IsVisible` is `true`
3. Evaluate visibility conditions on only those properties bound to the UI from then on

> If it is set to a constant true it is always visible

That could work too, and I'm open to that, though it feels a bit hacky to me (my comment earlier saying "rather than interpreting this expression as a constant"). If we go that road we would have to back out the CPS change that introduced `IsVisible` to the project model.

> I tried to make this change but as yet there doesn't seem to be a version of the Microsoft.VisualStudio.ProjectSystem.Query package that includes the IsVisible property.

I've asked the CPS folks for guidance here on Teams. cc @adrianvmsft

Reply:
Adrian pointed to a new package on Teams. I tested it out and ended up filing #7271 to show the approach I envisaged.